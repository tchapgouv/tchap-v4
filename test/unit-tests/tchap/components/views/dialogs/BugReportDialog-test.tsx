import * as React from "react";
import { RenderResult, fireEvent, render } from "@testing-library/react";

import BugReportDialog from "~linked-dep/matrix-react-sdk/src/components/views/dialogs/BugReportDialog";
import { flushPromises } from "~linked-dep/matrix-react-sdk/test/test-utils";
import SdkConfig from "~linked-dep/matrix-react-sdk/src/SdkConfig";

describe("<BugReportDialog>", () => {
    const bugReportUrl = "/bug/report/url";
    const userText = "halo i haz problem";
    const label = "mylabel";
    let sendBugReportSpy: jest.SpyInstance;

    beforeAll(() => {
        SdkConfig.reset(); // in case other tests didn't clean up
        SdkConfig.put({ bug_report_endpoint_url: bugReportUrl });
    });

    afterAll(function () {
        SdkConfig.reset(); // we touch the config, so clean up
    });

    beforeEach(() => {
        // Mocking trick to mock a default export : use requireActual.
        const actual = jest.requireActual("~linked-dep/matrix-react-sdk/src/rageshake/submit-rageshake");
        sendBugReportSpy = jest.spyOn(actual, "default").mockResolvedValue(bugReportUrl);
    });

    afterEach(() => {
        sendBugReportSpy.mockRestore();
    });

    const doSendLogs = async (component: RenderResult) => {
        // Note : component.getByX is preferable, and fails nicely if not found.
        // But querySelector is good enough and doesn't need to change the prod code.
        const button = component.container.querySelector(".mx_BugReportDialog_send_logs button");
        expect(button).toBeTruthy(); // button should be found
        fireEvent.click(button!);
        await flushPromises(); // needed to run through all the chained promises in BugReportDialog.onSubmit
    };

    it("should render as expected", async () => {
        const component = render(<BugReportDialog initialText={userText} label={label} onFinished={() => {}} />);
        expect(component).toMatchSnapshot();
    });

    it("sends the bug report", async () => {
        // userText passed as prop. We don't test the user actually filling the textbox.
        const component = render(<BugReportDialog initialText={userText} label={label} onFinished={() => {}} />);

        await doSendLogs(component);

        expect(sendBugReportSpy).toHaveBeenCalledWith(
            bugReportUrl,
            expect.objectContaining({
                progressCallback: expect.any(Function),
                labels: [label],
                userText: expect.stringContaining(userText),
                sendLogs: true,
            }),
        );
    });

    it("does not send the bug report if no userText", async () => {
        // No userText passed as prop.
        const component = render(<BugReportDialog initialText="" label={label} onFinished={() => {}} />);

        await doSendLogs(component);

        expect(sendBugReportSpy).not.toHaveBeenCalled();
        // Error should be displayed. We don't test the error text to allow flexibility if we change it.
        const error = component.container.querySelector(".error");
        expect(error).toBeTruthy(); // error should be found
    });

    it("adds voip fields in case of voip report", async () => {
        // We don't mock TchapUtils.isCurrentlyUsingBluetooth(), so it returns false.

        const component = render(
            <BugReportDialog initialText={userText} label="voip-feedback" onFinished={() => {}} />,
        );

        await doSendLogs(component);

        expect(sendBugReportSpy).toHaveBeenCalledWith(
            bugReportUrl,
            expect.objectContaining({
                labels: expect.arrayContaining(["voip-feedback"]),
                customFields: expect.objectContaining({ context: "voip", audio_input: "device" }),
            }),
        );
    });
});
