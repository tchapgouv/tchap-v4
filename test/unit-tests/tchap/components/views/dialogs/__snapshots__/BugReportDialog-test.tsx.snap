// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BugReportDialog> should render as expected 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        aria-describedby="mx_Dialog_content"
        aria-labelledby="mx_BaseDialog_title"
        class="mx_BugReportDialog mx_Dialog_fixedWidth"
        data-focus-lock-disabled="false"
        role="dialog"
      >
        <div
          class="mx_Dialog_header"
        >
          <h1
            class="mx_Heading_h3 mx_Dialog_title"
            id="mx_BaseDialog_title"
          >
            Submit debug logs to Tchap support team
          </h1>
        </div>
        <div
          class="mx_Dialog_content"
          id="mx_Dialog_content"
        >
          <p>
            <strong>
              Reminder: Your browser is unsupported, so your experience may be unpredictable.
            </strong>
          </p>
          <p>
            Debug logs contain application usage data including your username, the IDs or aliases of the rooms you have visited, which UI elements you last interacted with, and the usernames of other users. They do not contain messages.
          </p>
          <div
            class="mx_Field mx_Field_textarea mx_BugReportDialog_field_input"
          >
            <textarea
              id="mx_Field_1"
              label="Notes"
              placeholder="If there is additional context that would help in analysing the issue, such as what you were doing at the time, room IDs, user IDs, etc., please include those things here."
              rows="5"
              type="text"
            >
              halo i haz problem
            </textarea>
            <label
              for="mx_Field_1"
            >
              Notes
            </label>
          </div>
          <div
            class="mx_BugReportDialog_send_logs"
          >
            <div
              class="mx_Dialog_buttons"
            >
              <span
                class="mx_Dialog_buttons_row"
              >
                <button
                  class="mx_Dialog_primary"
                  data-testid="dialog-primary-button"
                  type="button"
                >
                  Send logs
                </button>
              </span>
            </div>
          </div>
          <div
            class="mx_BugReportDialog_download"
          >
            <p>
              Just want to get your own logs, without sharing them with the Tchap team?
            </p>
            <div
              class="mx_Dialog_buttons"
            >
              <span
                class="mx_Dialog_buttons_row"
              >
                <button
                  class="mx_Dialog_primary"
                  data-testid="dialog-primary-button"
                  type="button"
                >
                  Download logs
                </button>
              </span>
            </div>
          </div>
        </div>
        <div
          aria-label="Close dialog"
          class="mx_AccessibleButton mx_Dialog_cancelButton"
          role="button"
          tabindex="0"
        />
      </div>
      <div
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      aria-describedby="mx_Dialog_content"
      aria-labelledby="mx_BaseDialog_title"
      class="mx_BugReportDialog mx_Dialog_fixedWidth"
      data-focus-lock-disabled="false"
      role="dialog"
    >
      <div
        class="mx_Dialog_header"
      >
        <h1
          class="mx_Heading_h3 mx_Dialog_title"
          id="mx_BaseDialog_title"
        >
          Submit debug logs to Tchap support team
        </h1>
      </div>
      <div
        class="mx_Dialog_content"
        id="mx_Dialog_content"
      >
        <p>
          <strong>
            Reminder: Your browser is unsupported, so your experience may be unpredictable.
          </strong>
        </p>
        <p>
          Debug logs contain application usage data including your username, the IDs or aliases of the rooms you have visited, which UI elements you last interacted with, and the usernames of other users. They do not contain messages.
        </p>
        <div
          class="mx_Field mx_Field_textarea mx_BugReportDialog_field_input"
        >
          <textarea
            id="mx_Field_1"
            label="Notes"
            placeholder="If there is additional context that would help in analysing the issue, such as what you were doing at the time, room IDs, user IDs, etc., please include those things here."
            rows="5"
            type="text"
          >
            halo i haz problem
          </textarea>
          <label
            for="mx_Field_1"
          >
            Notes
          </label>
        </div>
        <div
          class="mx_BugReportDialog_send_logs"
        >
          <div
            class="mx_Dialog_buttons"
          >
            <span
              class="mx_Dialog_buttons_row"
            >
              <button
                class="mx_Dialog_primary"
                data-testid="dialog-primary-button"
                type="button"
              >
                Send logs
              </button>
            </span>
          </div>
        </div>
        <div
          class="mx_BugReportDialog_download"
        >
          <p>
            Just want to get your own logs, without sharing them with the Tchap team?
          </p>
          <div
            class="mx_Dialog_buttons"
          >
            <span
              class="mx_Dialog_buttons_row"
            >
              <button
                class="mx_Dialog_primary"
                data-testid="dialog-primary-button"
                type="button"
              >
                Download logs
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        aria-label="Close dialog"
        class="mx_AccessibleButton mx_Dialog_cancelButton"
        role="button"
        tabindex="0"
      />
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
